---
title: "String manipulation"
author: "NN"
date: "2021-10-25"
done: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(tidyr)
library(dplyr)
library(stringr)

```

> This little tutorial aims to make you familiar with some of the functions of the `stringr` package and a few regular expressions. 

## Strings and escape sequences in R

##### Write a sentence with escape sequences.
Try the sentence: `"It's the end of the world!" he said.\` . 
Assign the string to a variable and try as`print()`, `cat()` and `writeLines()`.

```{r}

mystring <- '"It\'s the end of the world!" he said.\\'
mystring
```

```{r}
cat(mystring)

writeLines(mystring)
```


## `stringr` functions 

We will be using the `words` data that is built into `stringr`. 
The data set is available to you if you load the package.


#### Words
```{r}
words
```
##### Select words that
1. Contain a `y`
```{r}
str_subset(words, "y")
```

2. Start with `y`
```{r}
str_subset(words, "^y")
```


3. Contain a `y` within the word

```{r}
str_subset(words, "[:alpha:][y][:alpha:]")
```

4. Extract the `y` and the previous character.

Note: Use the function `unique()` around the results to avoid printing many empty matches.

```{r}
tail(unique(str_extract(words, "[:alpha:][y]")),n=-1)
typeof(unique(str_extract(words, "[:alpha:][y]")))

```


5. Get the lengths of the first ten words

You can use `head(words, 10)` as a convenient way to access the elements of the `words` vector.
```{r}
```

> Missing

## Viral research 

Read the genome sequence of the Hepatitis D virus: [hepd.fasta](https://biostat2.uni.lu/practicals/data/hepd.fasta).
For now, just execute the following:

```{r}
hepd_genome <- readr::read_lines("https://biostat2.uni.lu/practicals/data/hepd.fasta")

hepd_genome


```

##### What is the length of the genome sequence? 

```{r}
nchar(hepd_genome)
```

> str_length, but OK

##### What is the sequence composition? How often does each character occur?

```{r}
uniqchars <- function(x) unique(strsplit(x, "")[[1]]) 
letters <- uniqchars(hepd_genome)
n <- length(letters)
n
letters[1]

count_let <- function(ch, n) c(letters[n], str_count(hepd_genome, letters[n])) 
#str_count(hepd_genome, letters[1])
for (i in c(1:n)){
  print(count_let(hepd_genome, i))
}


#seqs <- unique(str_extract_all(hepd_genome, #"[A-Z]{3}"))
#seqs


```

> This is not using the stringr function

##### Find motifs in the sequence using `str_locate()`. 

Find all matches of the sequence $ATG$ in the genome sequence. 

```{r}
str_locate_all(hepd_genome, "ATG")
```
 > Correct.

